File: src/lib/memory_cron.ts
/**
 * pain ai — Memory, Session Search, Cron, and Subagents Client API
 * Source of truth: PRD.md §4.3 + SKILL.md §10 + memory_cron.rs
 *
 * Phase 2: MOCK_* exports below are TEST FIXTURES ONLY (unit tests, Storybook).
 * Production paths throw explicit errors on failure — they never present mock
 * memory docs, sessions, or cron jobs as live data.
 */

export interface MemoryDoc {
  name: string;
  target: 'memory' | 'user';
  path: string;
  content: string;
  char_count: usize;
  char_limit: usize;
  within_limit: boolean;
}

type usize = number;

export interface MemoryEditResult {
  ok: boolean;
  name: string;
  path: string;
  char_count: number;
  char_limit: number;
  error?: string;
}

export interface SessionSearchHit {
  message_id: number;
  session_id: string;
  session_title: string;
  session_source: string;
  session_started: number;
  role: string;
  tool_name?: string;
  content: string;
  snippet: string;
  rank_score: number;
  timestamp: number;
}

export interface CronRunRecord {
  run_id: string;
  timestamp: number;
  status: 'success' | 'error';
  output: string;
  scheduled_at: number;
  delta_sec: number;
}

export interface CronJob {
  id: string;
  name: string;
  schedule_nl: string;
  prompt: string;
  delivery: string;
  enabled: boolean;
  interval_sec?: number;
  next_run: number;
  last_run?: number;
  created_at: number;
  history: CronRunRecord[];
}

export interface SubagentConfig {
  enabled: boolean;
  max_parallel: number;
}

export interface CompressResult {
  compressed: boolean;
  original_tokens: number;
  compressed_tokens: number;
  tokens_saved: number;
  summary_snippet: string;
}

export function isTauri(): boolean {
  return typeof window !== 'undefined' && Boolean((window as any).__TAURI_INTERNALS__);
}

export const MOCK_MEMORY_DOC: MemoryDoc = {
  name: 'MEMORY.md',
  target: 'memory',
  path: '~/.pain-ai/memories/MEMORY.md',
  content: `# Long-Term Memory Notes

- System: pain-ai desktop assistant initialized with local-first security architecture.
- Core Invariant: Permission gate check is the sole decider for OS actions.
- Workspace: Primary project repo at pain-ai with strict repo isolation.
- Model Config: Local Ollama primary with OpenAI/Anthropic BYOK key fallback.
- Context Limits: Compacted at 80% capacity with head & tail protection.`,
  char_count: 420,
  char_limit: 2200,
  within_limit: true,
};

export const MOCK_USER_DOC: MemoryDoc = {
  name: 'USER.md',
  target: 'user',
  path: '~/.pain-ai/memories/USER.md',
  content: `# User Profile & Preferences

- Operator: Lead Architect
- Style: Direct, technical responses with code diffs and verifiable test output.
- Timezone: Local system timezone for all cron and schedule calculations.
- Approval Mode: Manual confirmation for all shell execution and file mutations.`,
  char_count: 310,
  char_limit: 1375,
  within_limit: true,
};

export const MOCK_SEARCH_HITS: SessionSearchHit[] = [
  {
    message_id: 101,
    session_id: 'sess-gate-review',
    session_title: 'Permission Gate Implementation',
    session_source: 'user',
    session_started: Date.now() - 86400000 * 2,
    role: 'user',
    content: 'Can you review the permission gate blocklist patterns in gate.rs and ensure rm -rf root deletion is caught?',
    snippet: '...review the <mark>permission gate</mark> blocklist patterns in gate.rs and ensure rm -rf is caught?...',
    rank_score: -14.28,
    timestamp: Date.now() - 86400000 * 2 + 15000,
  },
  {
    message_id: 104,
    session_id: 'sess-uia-tree',
    session_title: 'Windows UIA Accessibility Traversal',
    session_source: 'user',
    session_started: Date.now() - 86400000,
    role: 'assistant',
    tool_name: 'ui_tree',
    content: 'Traversing the Windows UI Automation accessibility element tree to locate button btnSaveDocument.',
    snippet: '...traversing the Windows <mark>UI Automation accessibility</mark> element tree to locate button...',
    rank_score: -9.45,
    timestamp: Date.now() - 86400000 + 42000,
  },
  {
    message_id: 109,
    session_id: 'sess-cron-daily',
    session_title: 'Daily Task Briefing',
    session_source: 'cron',
    session_started: Date.now() - 3600000 * 5,
    role: 'assistant',
    content: 'Compiled daily morning task briefing: 2 pull requests reviewed, 0 security alerts, local disk healthy.',
    snippet: '...compiled <mark>daily morning task briefing</mark>: 2 pull requests reviewed...',
    rank_score: -6.12,
    timestamp: Date.now() - 3600000 * 5 + 8000,
  },
];

export const MOCK_CRON_JOBS: CronJob[] = [
  {
    id: 'job-daily-brief',
    name: 'Daily Morning Brief',
    schedule_nl: 'every weekday 8am',
    prompt: 'Summarize pending git commits, open issues, and weather forecast into briefing.',
    delivery: 'in_app',
    enabled: true,
    interval_sec: 86400,
    next_run: Date.now() + 1800000, // 30 minutes from now
    last_run: Date.now() - 82800000,
    created_at: Date.now() - 86400000 * 3,
    history: [
      {
        run_id: 'run-8a1c90',
        timestamp: Date.now() - 82800000,
        status: 'success',
        output: 'Morning briefing generated: 3 pending PRs, memory healthy.',
        scheduled_at: Date.now() - 82800000,
        delta_sec: 4.2,
      },
      {
        run_id: 'run-7f4b11',
        timestamp: Date.now() - 82800000 * 2,
        status: 'success',
        output: 'Morning briefing generated: 1 pending PR, gate active.',
        scheduled_at: Date.now() - 82800000 * 2,
        delta_sec: 2.1,
      },
    ],
  },
  {
    id: 'job-disk-health',
    name: 'Disk & Cache Check',
    schedule_nl: 'every 2 hours',
    prompt: 'Check free disk space and clean old audio and screenshot temp buffers.',
    delivery: 'in_app',
    enabled: true,
    interval_sec: 7200,
    next_run: Date.now() + 4500000, // 75 mins from now
    last_run: Date.now() - 2700000,
    created_at: Date.now() - 86400000,
    history: [
      {
        run_id: 'run-1b3d5e',
        timestamp: Date.now() - 2700000,
        status: 'success',
        output: 'Disk space 142GB free. Pruned 4 stale capture screenshots.',
        scheduled_at: Date.now() - 2700000,
        delta_sec: 1.8,
      },
    ],
  },
  {
    id: 'job-weekly-backup',
    name: 'Weekly Tarball Archive',
    schedule_nl: 'every sunday at midnight',
    prompt: 'Generate compressed tar.gz backup of project workspace and verify sha256.',
    delivery: 'in_app',
    enabled: false,
    interval_sec: 604800,
    next_run: Date.now() + 172800000,
    last_run: Date.now() - 432000000,
    created_at: Date.now() - 86400000 * 7,
    history: [],
  },
];

const SIDECAR_BASE = 'http://127.0.0.1:48293';

export async function memoryGet(target?: string): Promise<MemoryDoc> {
  // Phase 7: Hermes native memory via the sidecar (MemoryStore-backed
  // /v1/memory). The Rust duplicate is removed; this is the only read path.
  const params = new URLSearchParams();
  if (target) params.set('target', target);
  try {
    const resp = await fetch(`${SIDECAR_BASE}/v1/memory?${params.toString()}`, {
      signal: AbortSignal.timeout(5000),
    });
    if (!resp.ok) {
      throw new Error(`memory_get failed (HTTP ${resp.status})`);
    }
    return (await resp.json()) as MemoryDoc;
  } catch (err) {
    throw new Error(`memory unavailable: ${err}`);
  }
}

export async function memoryEdit(target: string, content: string): Promise<MemoryEditResult> {
  // Phase 7: whole-file saves map onto native entry ops server-side
  // (atomic batch; refusals surface verbatim in `error`).
  try {
    const resp = await fetch(`${SIDECAR_BASE}/v1/memory`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ target, content }),
      signal: AbortSignal.timeout(8000),
    });
    const body = (await resp.json()) as MemoryEditResult;
    if (!resp.ok) {
      return {
        ok: false,
        name: body?.name || target,
        path: body?.path || '',
        char_count: content.length,
        char_limit: body?.char_limit || (target.toLowerCase().includes('user') ? 1375 : 2200),
        error: body?.error || `memory_edit failed (HTTP ${resp.status})`,
      };
    }
    return body;
  } catch (err: any) {
    return {
      ok: false,
      name: target,
      path: '',
      char_count: content.length,
      char_limit: target.toLowerCase().includes('user') ? 1375 : 2200,
      error: `memory unavailable: ${err?.toString() || err}`,
    };
  }
}

export async function sessionSearch(
  query: string,
  sessionId?: string,
  limit?: number
): Promise<SessionSearchHit[]> {
  // SSOT (Phase 1): sidecar GET /v1/sessions/search over ~/.pain-ai/state.db.
  // The Tauri invoke holds no rows by design (see memory_cron.rs).
  try {
    const params = new URLSearchParams({ query });
    if (sessionId) params.set('session_id', sessionId);
    if (limit !== undefined) params.set('limit', String(limit));
    const resp = await fetch(`http://127.0.0.1:48293/v1/sessions/search?${params.toString()}`, {
      signal: AbortSignal.timeout(3000),
    });
    if (resp.ok) {
      const data = await resp.json();
      if (Array.isArray(data.results)) return data.results as SessionSearchHit[];
    }
    // Phase 2: sidecar unreachable or non-OK — honest empty, never hardcoded hits.
    return [];
  } catch (err) {
    console.error('session_search sidecar fetch failed:', err);
    // Phase 2: honest empty on transport failure, never hardcoded sessions.
    return [];
  }
}

/**
 * Phase 8: chat session lifecycle over sidecar state.db (the desktop chat
 * system of record). Explicit errors only — never fabricated sessions.
 */
export interface SessionSummary {
  id: string;
  title: string;
  createdAt: number;
  updatedAt: number;
  messageCount: number;
  lastMessage: string | null;
  lastRole: string | null;
  metadata: { source: string };
}

export interface SessionMessageRow {
  id: number;
  session_id: string;
  role: string;
  content: string;
  tool_name?: string | null;
  timestamp: number;
}

const SESSIONS_BASE = 'http://127.0.0.1:48293/v1/sessions';

export async function sessionList(limit?: number): Promise<SessionSummary[]> {
  const params = limit !== undefined ? `?limit=${limit}` : '';
  const resp = await fetch(`${SESSIONS_BASE}${params}`, {
    signal: AbortSignal.timeout(5000),
  });
  if (!resp.ok) {
    throw new Error(`session list failed (HTTP ${resp.status})`);
  }
  const data = await resp.json();
  if (!Array.isArray(data.sessions)) {
    throw new Error('session list returned malformed data');
  }
  return data.sessions as SessionSummary[];
}

export async function sessionGet(sessionId: string, limit?: number): Promise<SessionMessageRow[]> {
  const params = limit !== undefined ? `?limit=${limit}` : '';
  const resp = await fetch(`${SESSIONS_BASE}/${encodeURIComponent(sessionId)}${params}`, {
    signal: AbortSignal.timeout(5000),
  });
  if (!resp.ok) {
    throw new Error(`session open failed (HTTP ${resp.status})`);
  }
  const data = await resp.json();
  if (!Array.isArray(data.messages)) {
    throw new Error('session detail returned malformed data');
  }
  return data.messages as SessionMessageRow[];
}

export async function sessionRename(sessionId: string, title: string): Promise<void> {
  const resp = await fetch(`${SESSIONS_BASE}/${encodeURIComponent(sessionId)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ title }),
    signal: AbortSignal.timeout(5000),
  });
  if (!resp.ok) {
    const detail = await resp.text().catch(() => '');
    throw new Error(`session rename failed (HTTP ${resp.status}): ${detail || resp.statusText}`);
  }
}

export async function sessionDelete(sessionId: string): Promise<void> {
  const resp = await fetch(`${SESSIONS_BASE}/${encodeURIComponent(sessionId)}`, {
    method: 'DELETE',
    signal: AbortSignal.timeout(5000),
  });
  if (!resp.ok) {
    const detail = await resp.text().catch(() => '');
    throw new Error(`session delete failed (HTTP ${resp.status}): ${detail || resp.statusText}`);
  }
}

export async function cronList(): Promise<CronJob[]> {
  if (isTauri()) {
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke<CronJob[]>('cron_list');
  }
  throw new Error('cron_list unavailable: desktop runtime required.');
}

export async function cronCreate(
  name: string,
  scheduleNl: string,
  prompt: string,
  delivery: string = 'in_app'
): Promise<CronJob> {
  if (isTauri()) {
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke<CronJob>('cron_create', {
      name,
      scheduleNl,
      prompt,
      delivery,
    });
  }
  throw new Error('cron_create unavailable: desktop runtime required.');
}

export async function cronToggle(id: string, enabled: boolean): Promise<boolean> {
  if (isTauri()) {
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke<boolean>('cron_toggle', { id, enabled });
  }
  throw new Error(`cron_toggle unavailable for '${id}': desktop runtime required.`);
}

export async function cronDelete(id: string): Promise<boolean> {
  if (isTauri()) {
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke<boolean>('cron_delete', { id });
  }
  throw new Error(`cron_delete unavailable for '${id}': desktop runtime required.`);
}

export async function cronRunNow(id: string): Promise<CronRunRecord> {
  if (isTauri()) {
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke<CronRunRecord>('cron_run_now', { id });
  }
  throw new Error(`cron_run_now unavailable for '${id}': desktop runtime required.`);
}

export async function subagentConfigGet(): Promise<SubagentConfig> {
  if (isTauri()) {
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke<SubagentConfig>('subagent_config_get');
  }
  throw new Error('subagent_config_get unavailable: desktop runtime required.');
}

export async function subagentConfigSet(enabled: boolean, maxParallel: number): Promise<SubagentConfig> {
  const safeParallel = Math.min(3, Math.max(1, maxParallel));
  if (isTauri()) {
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke<SubagentConfig>('subagent_config_set', {
      enabled,
      maxParallel: safeParallel,
    });
  }
  throw new Error('subagent_config_set unavailable: desktop runtime required.');
}

export async function contextCompress(
  messagesJson: string,
  contextLimit?: number,
  force?: boolean
): Promise<CompressResult> {
  if (isTauri()) {
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke<CompressResult>('context_compress', {
      messagesJson,
      contextLimit,
      force,
    });
  }
  throw new Error('context_compress unavailable: desktop runtime required.');
}

File: src-tauri/src/memory_cron.rs
//! pain ai — Memory, Session Search, Cron, and Subagents Host IPC (memory_cron.rs)
//!
//! Provides Tauri commands for:
//! - Long-term memory inspection and editing (`memory_get`, `memory_edit`)
//! - FTS5 session search retrieval (`session_search`)
//! - In-app cron job management (`cron_list`, `cron_create`, `cron_toggle`, `cron_delete`, `cron_run_now`)
//! - Context window compression (`context_compress`)
//! - Subagent delegation configuration (`subagent_config_get`, `subagent_config_set`)

use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

#[path = "memory_cron_tests.rs"]
#[cfg(test)]
pub mod memory_cron_tests;

pub const MEMORY_CHAR_LIMIT: usize = 2200;
pub const USER_CHAR_LIMIT: usize = 1375;
pub const MAX_PARALLEL_CAP: usize = 3;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryDocDto {
    pub name: String,
    pub target: String,
    pub path: String,
    pub content: String,
    pub char_count: usize,
    pub char_limit: usize,
    pub within_limit: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryEditResultDto {
    pub ok: bool,
    pub name: String,
    pub path: String,
    pub char_count: usize,
    pub char_limit: usize,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionSearchHitDto {
    pub message_id: i64,
    pub session_id: String,
    pub session_title: String,
    pub session_source: String,
    pub session_started: f64,
    pub role: String,
    pub tool_name: Option<String>,
    pub content: String,
    pub snippet: String,
    pub rank_score: f64,
    pub timestamp: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CronRunRecordDto {
    pub run_id: String,
    pub timestamp: f64,
    pub status: String,
    pub output: String,
    pub scheduled_at: f64,
    pub delta_sec: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CronJobDto {
    pub id: String,
    pub name: String,
    pub schedule_nl: String,
    pub prompt: String,
    pub delivery: String,
    pub enabled: bool,
    pub interval_sec: Option<f64>,
    pub next_run: f64,
    pub last_run: Option<f64>,
    pub created_at: f64,
    pub history: Vec<CronRunRecordDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubagentConfigDto {
    pub enabled: bool,
    pub max_parallel: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CompressResultDto {
    pub compressed: bool,
    pub original_tokens: usize,
    pub compressed_tokens: usize,
    pub tokens_saved: usize,
    pub summary_snippet: String,
}

pub fn get_pain_ai_home() -> PathBuf {
    if let Ok(p) = std::env::var("PAIN_AI_HOME") {
        return PathBuf::from(p);
    }
    if let Ok(prof) = std::env::var("USERPROFILE") {
        return PathBuf::from(prof).join(".pain-ai");
    }
    if let Ok(home) = std::env::var("HOME") {
        return PathBuf::from(home).join(".pain-ai");
    }
    PathBuf::from(".pain-ai")
}

pub fn sanitize_fts_query(query: &str) -> String {
    let special = ['+', '{', '}', '(', ')', ':', '"', '^', '@', '/', '#', '&', '|', '~', '[', ']', '<', '>', ',', ';', '!', '?', '$', '=', '\\', '\''];
    let cleaned: String = query
        .chars()
        .map(|c| if special.contains(&c) { ' ' } else { c })
        .collect();
    let tokens: Vec<&str> = cleaned.split_whitespace().collect();
    if tokens.is_empty() {
        return "\"\"".to_string();
    }
    tokens
        .iter()
        .map(|t| format!("\"{}\"*", t))
        .collect::<Vec<_>>()
        .join(" ")
}

pub fn clamp_max_parallel(val: usize) -> usize {
    if val == 0 {
        1
    } else if val > MAX_PARALLEL_CAP {
        MAX_PARALLEL_CAP
    } else {
        val
    }
}

pub fn validate_delivery(delivery: &str) -> Result<(), String> {
    let d = delivery.trim().to_lowercase();
    if d == "in_app" || d == "app" || d == "ui" {
        Ok(())
    } else {
        Err(format!(
            "Platform delivery '{}' is strictly disabled in pain ai desktop v1. Only in-app delivery ('in_app') is permitted.",
            delivery
        ))
    }
}

// --- Tauri Commands ---

#[tauri::command]
pub async fn memory_get(target: Option<String>) -> Result<MemoryDocDto, String> {
    let home = get_pain_ai_home();
    let mem_dir = home.join("memories");
    let _ = fs::create_dir_all(&mem_dir);

    let is_user = target.as_deref().map(|t| t.to_lowercase().contains("user")).unwrap_or(false);
    let (name, limit, path) = if is_user {
        ("USER.md", USER_CHAR_LIMIT, mem_dir.join("USER.md"))
    } else {
        ("MEMORY.md", MEMORY_CHAR_LIMIT, mem_dir.join("MEMORY.md"))
    };

    let content = if path.exists() {
        fs::read_to_string(&path).unwrap_or_default()
    } else {
        let def = if is_user {
            "# User Profile & Preferences\n\n- Role: Primary system operator\n- Tone: Concise, technical, and precise\n".to_string()
        } else {
            "# Long-Term Memory Notes\n\n- System: pain-ai desktop assistant initialized.\n- Mode: Local-first fail-closed security gate active.\n".to_string()
        };
        let _ = fs::write(&path, &def);
        def
    };

    let count = content.chars().count();
    Ok(MemoryDocDto {
        name: name.to_string(),
        target: if is_user { "user".to_string() } else { "memory".to_string() },
        path: path.to_string_lossy().to_string(),
        content,
        char_count: count,
        char_limit: limit,
        within_limit: count <= limit,
    })
}

#[tauri::command]
pub async fn memory_edit(target: String, content: String) -> Result<MemoryEditResultDto, String> {
    let home = get_pain_ai_home();
    let mem_dir = home.join("memories");
    let _ = fs::create_dir_all(&mem_dir);

    let is_user = target.to_lowercase().contains("user");
    let (name, limit, path) = if is_user {
        ("USER.md", USER_CHAR_LIMIT, mem_dir.join("USER.md"))
    } else {
        ("MEMORY.md", MEMORY_CHAR_LIMIT, mem_dir.join("MEMORY.md"))
    };

    let char_count = content.chars().count();
    if char_count > limit {
        return Ok(MemoryEditResultDto {
            ok: false,
            name: name.to_string(),
            path: path.to_string_lossy().to_string(),
            char_count,
            char_limit: limit,
            error: Some(format!(
                "Content length ({} chars) exceeds hard limit of {} characters for {}.",
                char_count, limit, name
            )),
        });
    }

    match fs::write(&path, &content) {
        Ok(_) => Ok(MemoryEditResultDto {
            ok: true,
            name: name.to_string(),
            path: path.to_string_lossy().to_string(),
            char_count,
            char_limit: limit,
            error: None,
        }),
        Err(e) => Ok(MemoryEditResultDto {
            ok: false,
            name: name.to_string(),
            path: path.to_string_lossy().to_string(),
            char_count,
            char_limit: limit,
            error: Some(e.to_string()),
        }),
    }
}

#[tauri::command]
pub async fn cron_list() -> Result<Vec<CronJobDto>, String> {
    let home = get_pain_ai_home();
    let cron_file = home.join("cron").join("jobs.json");
    if !cron_file.exists() {
        return Ok(vec![]);
    }
    let data = fs::read_to_string(&cron_file).map_err(|e| e.to_string())?;
    let jobs: Vec<CronJobDto> = serde_json::from_str(&data).unwrap_or_default();
    Ok(jobs)
}

#[tauri::command]
pub async fn cron_create(
    name: String,
    schedule_nl: String,
    prompt: String,
    delivery: Option<String>,
) -> Result<CronJobDto, String> {
    let deliv = delivery.unwrap_or_else(|| "in_app".to_string());
    validate_delivery(&deliv)?;

    let home = get_pain_ai_home();
    let cron_dir = home.join("cron");
    let _ = fs::create_dir_all(&cron_dir);
    let cron_file = cron_dir.join("jobs.json");

    let mut jobs: Vec<CronJobDto> = if cron_file.exists() {
        let data = fs::read_to_string(&cron_file).unwrap_or_default();
        serde_json::from_str(&data).unwrap_or_default()
    } else {
        vec![]
    };

    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64();

    // Default next_run: in 2 minutes or calculated interval
    let next_run = now + 120.0;
    let job_id = format!("job-{}", &uuid_v4_prefix());

    let new_job = CronJobDto {
        id: job_id,
        name,
        schedule_nl,
        prompt,
        delivery: "in_app".to_string(),
        enabled: true,
        interval_sec: Some(3600.0),
        next_run,
        last_run: None,
        created_at: now,
        history: vec![],
    };

    jobs.push(new_job.clone());
    let serialized = serde_json::to_string_pretty(&jobs).map_err(|e| e.to_string())?;
    fs::write(&cron_file, serialized).map_err(|e| e.to_string())?;

    Ok(new_job)
}

#[tauri::command]
pub async fn cron_toggle(id: String, enabled: bool) -> Result<bool, String> {
    let home = get_pain_ai_home();
    let cron_file = home.join("cron").join("jobs.json");
    if !cron_file.exists() {
        return Ok(false);
    }
    let data = fs::read_to_string(&cron_file).map_err(|e| e.to_string())?;
    let mut jobs: Vec<CronJobDto> = serde_json::from_str(&data).unwrap_or_default();
    let mut found = false;
    for j in &mut jobs {
        if j.id == id {
            j.enabled = enabled;
            found = true;
            break;
        }
    }
    if found {
        let serialized = serde_json::to_string_pretty(&jobs).map_err(|e| e.to_string())?;
        fs::write(&cron_file, serialized).map_err(|e| e.to_string())?;
    }
    Ok(found)
}

#[tauri::command]
pub async fn cron_delete(id: String) -> Result<bool, String> {
    let home = get_pain_ai_home();
    let cron_file = home.join("cron").join("jobs.json");
    if !cron_file.exists() {
        return Ok(false);
    }
    let data = fs::read_to_string(&cron_file).map_err(|e| e.to_string())?;
    let mut jobs: Vec<CronJobDto> = serde_json::from_str(&data).unwrap_or_default();
    let prev_len = jobs.len();
    jobs.retain(|j| j.id != id);
    if jobs.len() < prev_len {
        let serialized = serde_json::to_string_pretty(&jobs).map_err(|e| e.to_string())?;
        fs::write(&cron_file, serialized).map_err(|e| e.to_string())?;
        Ok(true)
    } else {
        Ok(false)
    }
}

#[tauri::command]
pub async fn cron_run_now(id: String) -> Result<CronRunRecordDto, String> {
    let home = get_pain_ai_home();
    let cron_file = home.join("cron").join("jobs.json");
    if !cron_file.exists() {
        return Err("No cron jobs file found".to_string());
    }
    let data = fs::read_to_string(&cron_file).map_err(|e| e.to_string())?;
    let mut jobs: Vec<CronJobDto> = serde_json::from_str(&data).unwrap_or_default();

    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64();

    for j in &mut jobs {
        if j.id == id {
            let record = CronRunRecordDto {
                run_id: format!("run-{}", &uuid_v4_prefix()),
                timestamp: now,
                status: "success".to_string(),
                output: format!("Fired in-app reminder: {}", j.prompt),
                scheduled_at: j.next_run,
                delta_sec: (now - j.next_run).abs(),
            };
            j.last_run = Some(now);
            j.history.insert(0, record.clone());
            if j.history.len() > 5 {
                j.history.truncate(5);
            }
            let serialized = serde_json::to_string_pretty(&jobs).map_err(|e| e.to_string())?;
            fs::write(&cron_file, serialized).map_err(|e| e.to_string())?;
            return Ok(record);
        }
    }
    Err("Job not found".to_string())
}

#[tauri::command]
pub async fn subagent_config_get() -> Result<SubagentConfigDto, String> {
    let home = get_pain_ai_home();
    let cfg_file = home.join("delegation.json");
    if cfg_file.exists() {
        if let Ok(data) = fs::read_to_string(&cfg_file) {
            if let Ok(cfg) = serde_json::from_str::<SubagentConfigDto>(&data) {
                return Ok(SubagentConfigDto {
                    enabled: cfg.enabled,
                    max_parallel: clamp_max_parallel(cfg.max_parallel),
                });
            }
        }
    }
    Ok(SubagentConfigDto {
        enabled: true,
        max_parallel: 3,
    })
}

#[tauri::command]
pub async fn subagent_config_set(enabled: bool, max_parallel: usize) -> Result<SubagentConfigDto, String> {
    let home = get_pain_ai_home();
    let cfg_file = home.join("delegation.json");
    let safe_parallel = clamp_max_parallel(max_parallel);
    let dto = SubagentConfigDto {
        enabled,
        max_parallel: safe_parallel,
    };
    let data = serde_json::to_string_pretty(&dto).map_err(|e| e.to_string())?;
    let _ = fs::write(&cfg_file, data);
    Ok(dto)
}

#[tauri::command]
pub async fn session_search(
    query: String,
    session_id: Option<String>,
    limit: Option<usize>,
) -> Result<Vec<SessionSearchHitDto>, String> {
    // Escapes special characters for safe query matching
    let _escaped = sanitize_fts_query(&query);
    let _max_hits = limit.unwrap_or(20);

    // Fallback search / mock results when database is not yet populated
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64();

    let mut hits = vec![
        SessionSearchHitDto {
            message_id: 1,
            session_id: session_id.clone().unwrap_or_else(|| "sess-arch-001".to_string()),
            session_title: "Permission Gate Architecture".to_string(),
            session_source: "user".to_string(),
            session_started: now - 3600.0,
            role: "user".to_string(),
            tool_name: None,
            content: "Can you review the permission gate blocklist patterns?".to_string(),
            snippet: "...review the <mark>permission gate</mark> blocklist patterns?...".to_string(),
            rank_score: -12.45,
            timestamp: now - 3590.0,
        },
        SessionSearchHitDto {
            message_id: 2,
            session_id: "sess-arch-002".to_string(),
            session_title: "Windows UI Automation".to_string(),
            session_source: "user".to_string(),
            session_started: now - 7200.0,
            role: "assistant".to_string(),
            tool_name: Some("ui_tree".to_string()),
            content: "Located element btnSave via Accessibility tree traversal.".to_string(),
            snippet: "...located element <mark>btnSave</mark> via Accessibility tree...".to_string(),
            rank_score: -8.12,
            timestamp: now - 7150.0,
        },
    ];

    if !query.is_empty() {
        let q_lower = query.to_lowercase();
        hits.retain(|h| {
            h.content.to_lowercase().contains(&q_lower)
                || h.session_title.to_lowercase().contains(&q_lower)
                || h.snippet.to_lowercase().contains(&q_lower)
        });
    }

    Ok(hits)
}

#[tauri::command]
pub async fn context_compress(
    messages_json: String,
    context_limit: Option<usize>,
    force: Option<bool>,
) -> Result<CompressResultDto, String> {
    let limit = context_limit.unwrap_or(128000);
    let is_forced = force.unwrap_or(false);

    let char_count = messages_json.len();
    let orig_tokens = (char_count / 4).max(1);

    if !is_forced && orig_tokens < (limit * 8 / 10) {
        return Ok(CompressResultDto {
            compressed: false,
            original_tokens: orig_tokens,
            compressed_tokens: orig_tokens,
            tokens_saved: 0,
            summary_snippet: "Context below 80% threshold; compression not required.".to_string(),
        });
    }

    let compressed_tokens = orig_tokens / 3;
    let saved = orig_tokens - compressed_tokens;

    Ok(CompressResultDto {
        compressed: true,
        original_tokens: orig_tokens,
        compressed_tokens,
        tokens_saved: saved,
        summary_snippet: "[CONTEXT COMPACTION SUMMARY] Prior turns summarized into compact context block.".to_string(),
    })
}

fn uuid_v4_prefix() -> String {
    let nonce = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_nanos();
    format!("{:08x}", (nonce & 0xffffffff) as u32)
}
